### PR TITLE
Add: openvasd: encryption of credentials and results

### PR DESCRIPTION
**Pull request by nichtsfrei**
To allow easier handling of sensitive data the direct usage of a hash
map is replaced with a storage interface and field encryption is added.

This is done by introducing interfaces:
- to get the progress of a scan
- store scans
- append results

instead of using a hashmap of a progress struct. This allows us to
exchange the storage from inmemory to something like a relational
database.

Additionally a crypt interface with a chacha20 implementation is
introduces to enable encryption of sensitive fields like passwords or
results.

[SC-900](https://jira.greenbone.net/browse/SC-900)

### CODE REVIEW

File: rust/openvasd/src/controller/entry.rs [hunk 1]
```diff
@@ -134,165 +107,128 @@ where
...
     }
+    tracing::debug!("{} {}", req.method(), kp);
 
     match (req.method(), kp) {
-        (&Method::HEAD, _) => Ok(ctx.response.empty(hyper::StatusCode::OK)),
         (&Method::POST, Scans(None)) => {
             match crate::request::json_request::<models::Scan>(&ctx.response, req).await {
                 Ok(mut scan) => {
-                    response_blocking(move || {
+                    if scan.scan_id.is_none() {
```
Comment (jjnicola) [inline rust/openvasd/src/controller/entry.rs:116]: Why is this `if` statement introduced again? In a previous PR we removed and even changed the API to not accept an scan_id for creating a scan

File: rust/openvasd/src/controller/entry.rs [hunk 2]
```diff
@@ -68,59 +68,32 @@ impl Display for KnownPaths {
...
 {
     use KnownPaths::*;
+    // on head requests we just return an empty response without checking the api key
+
+    if req.method() == Method::HEAD {
+        return Ok(ctx.response.empty(hyper::StatusCode::OK));
+    }
     let kp = KnownPaths::from_path(req.uri().path());
-    tracing::debug!("{} {}", req.method(), kp);
+    tracing::debug!("{} {}:{:?}", req.method(), kp, req.uri().query());
```
Comment (Kraemii) [inline rust/openvasd/src/controller/entry.rs:96]: I would suggest to put the logging before the handling of the HEAD request

File: rust/openvasd/src/controller/entry.rs [hunk 3]
```diff
@@ -134,165 +107,128 @@ where
             }
         }
     }
+    tracing::debug!("{} {}", req.method(), kp);
```
Comment (Kraemii) [inline rust/openvasd/src/controller/entry.rs:110]: What is the second log for? These information are already contained in the log before

File: rust/openvasd/src/controller/entry.rs (hunk 1 above)
Comment (Kraemii) [inline rust/openvasd/src/controller/entry.rs:116]: The scan ID should always be generated

**Review by jjnicola — CHANGES_REQUESTED** [review]

File: rust/openvasd/src/controller/entry.rs [hunk 4]
```diff
@@ -134,165 +107,128 @@ where
...
+                        use models::Phase::*;
+                        let expected = &[Stored, Stopped, Failed, Succeeded];
+                        Ok(ctx.response.not_accepted(&status.status, expected))
+                    } else {
+                        match ctx.scanner.start_scan(scan).await {
+                            Ok(_) => {
+                                status.status = models::Phase::Requested;
+                                match ctx.db.update_status(&id, status).await {
+                                    Ok(_) => Ok(ctx.response.no_content()),
                                     Err(e) => Ok(ctx.response.internal_server_error(&e)),
```
Comment (Kraemii) [inline rust/openvasd/src/controller/entry.rs:145]: I think we also have to stop the scan at this point, as
```rust
ctx.scanner.start_scan(scan).await
```
was successfully executed

File: rust/openvasd/src/controller/entry.rs (hunk 1 above)
Comment (nichtsfrei) [reply rust/openvasd/src/controller/entry.rs:116]: rebase error, going to remove it.

**Review by jjnicola — APPROVED** [review]

**Pull request merged**